### PR TITLE
Enabled internal-link in ckeditor

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | none
| Related issues/PRs | https://github.com/sulu/sulu-standard/pull/686 , https://github.com/massiveart/husky/pull/643 based on https://github.com/sulu/sulu/pull/2338
| License | MIT
| Documentation PR | https://github.com/sulu/sulu-docs/pull/224

#### What's in this PR?

This PR adds a ckeditor plugin which enables to add declarative links to other pages.

#### To Do

- [x] Create a documentation PR
- [x] ckeditor plugin
- [x] html replace of special tags with links to pages
- [x] preview
- [x] sulu-standard single-internal-link
- [x] translations
- [x] icon
- [x] extract review PR

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/Markup/LinkTag.php [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+use Sulu\Component\Content\Repository\Content;
+use Sulu\Component\Content\Repository\ContentRepositoryInterface;
+use Sulu\Component\Content\Repository\Mapping\MappingBuilder;
+use Sulu\Component\Webspace\Manager\WebspaceManagerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Extends the sulu markup with the "sulu:link" tag.
+ */
+class LinkTag implements TagInterface
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Markup/LinkTag.php]: missing testcase

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/ckeditor/internal-link.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                        ]);
+                    }
+                });
+
+                editor.ui.addButton(
+                    'InternalLink',
+                    {
+                        label: sandbox.translate('public.internal-link'),
+                        command: 'internalLinkDialog',
+                        // TODO icon
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/ckeditor/internal-link.js]: change icon

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/form.html [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                   class="form-element"
+                   data-mapper-property="altTitle"/>
+        </div>
+    </div>
+
+    <div class="loader"></div>
+
+    <div class="form-group grid-row target-container" style="display: none;">
+        <label for="internal-link-target-button" class="required">
+            <%= translations.target %>
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/form.html]: translations

**Comment by wachterjohannes** [comment]
this feature currently dont work for preview updates. but it should be working with the preview refactoring https://github.com/sulu/sulu/pull/2299

**Comment by wachterjohannes** [comment]
the `sulu:link` tag can also be used by single-internal-link and internal-link. i will update the docs and standard theme.

**Comment by wachterjohannes** [comment]
this PR is ready to review. but the icon should be replaced before merging.

**Comment by wachterjohannes** [comment]
how does this behave in all the other bundles (without webspace)

**Comment by wachterjohannes** [comment]
i have tested it with snippets and it works there (:

**Comment by alexander-schranz** [comment]
did you test this with custom editor configs?

an example used mostly on our websites:

```
<property name="description" type="text_editor">
    <meta>
        <title lang="de">Beschreibung</title>
        <title lang="en">Description</title>
    </meta>

    <params>
        <param name="link" value="true"/>
        <param name="table" value="true"/>
        <param name="paste_from_word" value="true"/>
        <param name="extra_allowed_content" value="img(*)[*]; span(*)[*]; div(*)[*]; iframe(*)[*]; script(*)[*]" />
    </params>
</property>
```

I remember that their were some issues here where also people did overwrite the ckeditor settings. So we should keep that in mind and add some lines to the `UPGRADE`.

**Comment by wachterjohannes** [comment]
@alexander-schranz i have tested it with the same config. the param `extra_allowed_content` will be removed in one of the next PR when "enhancing" the text_editor. in the future we will allow all the content and "restrict" only the toolbar over roles config.

**Comment by alexander-schranz** [comment]
@wachterjohannes yeah sounds good I just want that this need an entry in the `Upgrade`.

**Comment by wachterjohannes** [comment]
ill be added when we remove this optinos.

**Comment by wachterjohannes** [comment]
the last commit should be extracted into own PR

File: src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php [hunk 4]
```diff
@@ -155,7 +155,7 @@ public function render($object, $id, $webspaceKey, $locale, $partial = false)
         $this->eventDispatcher->dispatch(Events::PRE_RENDER, new PreRenderEvent($attributes));
 
         try {
-            $response = $this->httpKernel->handle($request, HttpKernelInterface::SUB_REQUEST, false);
+            $response = $this->httpKernel->handle($request, HttpKernelInterface::MASTER_REQUEST, false);
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/PreviewBundle/Preview/Renderer/PreviewRenderer.php]: This is necessary to ensure all listener will be called (like on the website).

File: src/Sulu/Component/Content/Repository/ContentRepository.php [hunk 5]
```diff
@@ -583,13 +580,10 @@ private function resolveContent(
         Row $row,
         $locale,
         $locales,
-        $webspaceKey,
```
Comment (wachterjohannes) [inline src/Sulu/Component/Content/Repository/ContentRepository.php]: Webspace-Key can be extracted from the path. this avoids some problems when using multiple webspaces in the same query.

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/form.html [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        </div>
+
+        <div class="form-group grid-col-6">
+            <label for="internal-link-target"><%= translations.target %></label>
+            <select id="internal-link-target"
+                    name="target"
+                    class="form-element"
+                    data-mapper-property="target">
+                <option value=""></option>
+                <option value="_blank"><%= translations.targetBlank%></option>
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/form.html]: space before `%>`

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+                href: 'content.ckeditor.internal-link.href',
+                target: 'content.ckeditor.internal-link.target',
+                targetBlank: 'content.ckeditor.internal-link.target-blank',
+                targetSelf: 'content.ckeditor.internal-link.target-self',
+                internalLink: 'content.ckeditor.internal-link'
+            }
+        },
+
+        initialize: function() {
+            this.initDialog();
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js]: No abbreviations

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+                ]
+            ).then(function() {
+                if (!this.options.link.href) {
+                    this.showHrefInput();
+
+                    return;
+                }
+
+                this.sandbox.once('husky.column-navigation.smart-content-internal-link.loaded', function() {
+                    this.sandbox.emit('husky.column-navigation.smart-content-internal-link.get-breadcrumb', function(breadcrumb) {
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js]: What about the `smart-content` in the previous two lines?

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+                        if (breadcrumb.length === 0) {
+                            this.showHrefInput();
+
+                            return;
+                        }
+
+                        this.setHref(breadcrumb[breadcrumb.length - 1]);
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js]: Do we have to load that via the breadcrumbs? Isn't there a more elegant version?

Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js]: the column-navigation knows the breadcrumb without any additional request. should i add there a event to get the current item?

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js (hunk 8 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js]: i will make the prefix configurable.

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js (hunk 9 above)
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/ckeditor-internal-link/main.js]: I guess it is ok for now.

**Comment by danrot** [comment]
The icons look a bit blurry on a retina screen:
image

**Comment by danrot** [comment]
The overlay can be saved like this:
image

Then there is no link shown on the website.

**Comment by danrot** [comment]
If I start one of the text editors on click, and hit the return button afterwards, I get the following javascript error, and the column navigation is not loading:

```
husky.js:28268 aura_1: Error caught in listener 'sulu.router.navigate', called with arguments:  ["content/contents/sulu_io/en"] 
Error: Cannot read property 'getFrame' of undefined TypeError: Cannot read property 'getFrame' of undefined
```

**Pull request merged**